src/operator/filter_map.rs

```diff
@@ -1,145 +1,487 @@
 
 use crate::{
+    algebra::MonoidValue,
     circuit::{
         operator_traits::{Operator, UnaryOperator},
         Circuit, Scope, Stream,
     },
-    trace::{Batch, BatchReader, Cursor},
+    trace::{cursor::Cursor, Batch, BatchReader, Builder},
     OrdIndexedZSet, OrdZSet,
 };
 use std::{borrow::Cow, marker::PhantomData};
 
-impl<P, CI> Stream<Circuit<P>, CI>
+/// This trait abstracts away a stream of records that can be filtered
+/// and transformed on a record-by-record basis.
+///
+/// The notion of a record is determined by each implementer and can be either
+/// a `(key, value)` tuple for a stream of key/value pairs or just `key`
+/// for a stream of singleton values.
+///
+/// # Background
+///
+/// DBSP represents relational data using the [`Batch`](`crate::trace::Batch`)
+/// trait. A batch is conceptually a set of `(key, value, weight)` tuples.  When
+/// processing batches, we often need to filter and/or transform tuples one at a
+/// time with a user-provided closure, e.g., we may want to create a batch with
+/// only the tuples that satisfy a predicate of the form `Fn(K, V) -> bool`.
+///
+/// In practice we often work with specialized implementations of `Batch` where
+/// the value type is `()`, e.g., [`OrdZSet`](`crate::OrdZSet`).  We call such
+/// batches **non-indexed batches**, in contrast to **indexed batches** like
+/// [`OrdIndexedZSet`](`crate::OrdIndexedZSet`), which support arbitrary `value`
+/// types.  When filtering or transforming non-indexed batches we want to ignore
+/// the value and work with keys only, e.g., to write filtering predicates of
+/// the form `Fn(K) -> bool` rather than `Fn(K, ()) -> bool`.
+///
+/// This trait enables both use cases by allowing the implementer to define
+/// their own record type, which can be either `(K, V)` or `K`.
+///
+/// # Naming convention
+///
+/// This trait uses `_index` and `_generic` suffixes to declare variations of
+/// basic operations, e.g., `map`, `map_index`, `map_generic`,
+/// `map_index_generic`.
+///
+/// ## `_generic` suffix
+///
+/// Most transformations in this trait can assemble their outputs into any
+/// collection type that implements the [`Batch`](`crate::trace::Batch`) trait.
+/// In practice, we typically use [`OrdIndexedZSet`](`crate::OrdIndexedZSet`)
+/// for indexed batches and [`OrdZSet`](`crate::OrdZSet`) for non-indexed
+/// batches.  Methods without the `_generic` suffix return these concrete types,
+/// eliminating the need to type-annotate each invocation, while `_generic`
+/// methods can be used to return arbitrary custom `Batch` implementations.
+///
+/// ## `_index` suffix
+///
+/// Methods without the `_index()` suffix return non-indexed batches.
+/// `<method>_index()` methods combine the effects of `<method>()` and
+/// [`index()`](`crate::Stream::index`), e.g., `stream.map_index(closure)` is
+/// functionally equivalent, but more efficient than, `stream.map(closure).
+/// index()`.
+pub trait FilterMap<C> {
```

Comment:
Shouldn't pretty much all of these take a KV pair? Otherwise you can't manipulate the value or filter based off of it

Reply:
That's the whole point of this trait.  You want KV pairs for IndexedZSet's and just K for regular ZSet's.  The `Self::Item` type allows both.